AmplifyPlugins/Core/AWSPluginsCore/Model/Support/Model+GraphQL.swift

```diff
@@ -18,76 +18,129 @@ extension Model {
             case .model:
-                // For Models, append the model name in front in case a targetName is not provided
-                // e.g. "comment" + "PostId"
-                var fieldName = modelName.camelCased() + name
-                if case let .belongsTo(_, targetName) = field.association {
-                    fieldName = targetName ?? fieldName
-                }
-                if let modelValue = value as? Model {
-                    input[fieldName] = modelValue.id
-
-                } else if let value = value as? [String: JSONValue],
-                          case .string(let primaryKeyValue) = value[modelSchema.primaryKey.name] {
-                    input[fieldName] = primaryKeyValue
-                } else {
-                    input[fieldName] = nil
-                }
+                let fieldName = getFieldNameForAssociatedModels(modelField: modelField) ?? modelField.name
+                input[fieldName] = getModelId(from: value, modelSchema: modelSchema)
             case .embedded, .embeddedCollection:
                 if let encodable = value as? Encodable {
                     let jsonEncoder = JSONEncoder(dateEncodingStrategy: ModelDateFormatting.encodingStrategy)
                     do {
                         let data = try jsonEncoder.encode(encodable.eraseToAnyEncodable())
-                        input[name] = try JSONSerialization.jsonObject(with: data)
+                        input[modelField.name] = try JSONSerialization.jsonObject(with: data)
                     } catch {
                         preconditionFailure("Could not turn into json object from \(value)")
                     }
                 }
+            case .string:
+                // The input may contain the same keys from `.model` case when the `getFieldNameForAssociatedModels`
+                // returns the same value as the `modelField.name`. If this is the case, let the `.model` case take
+                // precedent over the explicit string field on the Model by ignoring the value that is about to be added
+                if !input.keys.contains(modelField.name) {
+                    input[modelField.name] = value
+                }
             default:
-                input[name] = value
+                input[modelField.name] = value
             }
         }
+
+        return fixHasOneAssociationsWithExplicitFieldOnModel(input, modelSchema: modelSchema)
+    }
+
+    // This is to account for Models with an explicit field on the Model along with an object representing the hasOne
+    // association to another model. See https://github.com/aws-amplify/amplify-ios/issues/920 for more details.
+    // When the associated object is `nil`, remove the key from the GraphQL input to prevent runtime failures.
+    // When the associated object is found, take this value over the explicit field's value by replacing the correct
+    // entry for the field name of the associated model.
```

Comment:
I'm not necessarily disagreeing, but why is it right to take the object rather than the specified field? How do we let the customer know which one will be used?

Reply:
If they do any of the below, the correct team identifier will be used
```
let project = Project2(teamID: team.id, team: team)
let project = Project2(teamID: team.id) 
// note `teamID` is always required
```
So the decision to take the object over the specified field only matters when the developer sets their values differently, ie.
```
let project = Project2(teamID: "nonExistentTeamID", team: team)
```

If the developer started off with
```
let project = Project2(teamID: "nonExistentTeamID")
```
This will be persisted but the association points to an non existent team. They'll eventually end up with one of the three scenarios once they get working code, by persisting/retrieving the team object before using it in the project object
```
let project = Project2(teamID: team.id) // 1. they updated their code to use team.id, and it works without passing in an optional `team`
let project = Project2(teamID: team.id, team: team) // 2. they updated it in both places, because it's confusing where to add it to
let project = Project2(teamID: "nonExistentTeamID", team: team) // 3. they only passed it into `team` parameter, but it starts working. I'm guessing the teamID here is gibberish
```
Another thought I had for this was thinking that we will omit the specific field in the future with a breaking codegen change, so that our documentation always promotes the use of passing the object (since belongsTo associations already omit the specific field). So where it's not omitted, we would continue to document it as `let project = Project2(teamID: team.id, team: team)` to communicate that developers should work with the model, while satisfying requirement that `teamID` is a required field